x/programs/wasmlanche/src/memory.rs

```diff
@@ -13,22 +13,57 @@ use alloc::{
     alloc::{alloc as allocate, dealloc as deallocate, handle_alloc_error, Layout},
     vec::Vec,
 };
-use core::{
-    cell::{LazyCell, RefCell},
-    mem::ManuallyDrop,
-    ops::Deref,
-    slice,
-};
-use hashbrown::HashMap;
+use core::{mem::ManuallyDrop, ops::Deref, slice};
+
+/// Safety:
+/// Only safe to use in a single-threaded environment
+mod allocations {
+    use cfg_if::cfg_if;
+    use core::cell::RefCell;
+    use hashbrown::HashMap;
+
+    type LenMap = HashMap<*const u8, usize>;
+
+    // Map of pointer to the length of its content on the heap
+    cfg_if! {
+        if #[cfg(not(any(feature = "build", feature = "debug", test, target_arch = "wasm32")))] {
+            use core::cell::{LazyCell, Ref, RefMut};
+            static mut ALLOCATIONS: LazyCell<RefCell<LenMap>> = LazyCell::new(|| RefCell::new(LenMap::new()));
+
+            fn borrow<F: FnOnce(Ref<'_, LenMap>) -> R, R>(f: F) -> R {
+                f(unsafe { ALLOCATIONS.borrow() })
+            }
+
+            fn borrow_mut<F: FnOnce(&mut RefMut<'_, LenMap>) -> R, R>(f: F) -> R {
+                f(unsafe { &mut ALLOCATIONS.borrow_mut() })
+            }
+        } else {
+            std::thread_local! {
+                static ALLOCATIONS: RefCell<LenMap> = RefCell::new(HashMap::new());
+            }
+
+            fn borrow<F: FnOnce(&LenMap) -> R, R>(f: F) -> R {
+                ALLOCATIONS.with_borrow(f)
+            }
+
+            fn borrow_mut<F: FnOnce(&mut LenMap) -> R, R>(f: F) -> R {
+                ALLOCATIONS.with_borrow_mut(f)
+            }
```

Comment:
The secret sauce is in the `borrow` and `borrow_mut` functions. The business logic is the *exact* same in both cases (of the `cfg_if`); it just acts on different types (a `&HashMap<*const u8, size>>>` vs a `Ref<'_, HashMap<*const u8, usize>>` and the mutable equivalents).